Use select entities for Litter-Robot 4 night light functionality

natekspencer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Use select entity instead of a switch entity for night light mode on Litter-Robot 4 units.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The Litter-Robot 4 has a more enriched feature set for night light mode configuration. This exposes that functionality.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

natekspencer committed:
Use select entities for Litter-Robot 4 night light functionality

homeassistant commented:
Hey there @tkdrob, mind taking a look at this pull request as it has been labeled with an integration (`litterrobot`) you are listed as a [code owner](https://github.com/home-assistant/core/blob/dev/CODEOWNERS#L627) for? Thanks!

tkdrob commented on homeassistant/components/litterrobot/select.py:
```diff
@@ -42,26 +42,71 @@ class RobotSelectEntityDescription(
+            select_fn=lambda robot, option: robot.set_night_light_brightness(
+                NightLightLevel[option.upper()]
+            ),
+            icon_fn=lambda robot: "mdi:lightbulb-question"
+            if (level := robot.night_light_level) is None
+            else "mdi:lightbulb-on"
+            if level == NightLightLevel.HIGH
+            else "mdi:lightbulb-on-70"
+            if level == NightLightLevel.MEDIUM
+            else "mdi:lightbulb-on-30",
```
Should this be a separate method? Ternary conditional operator spanning multiple lines may not be preferred.

natekspencer replied:
Yeah, stupid brain fart moment. I blame Monday.

tkdrob commented on homeassistant/components/litterrobot/select.py:
```diff
@@ -42,26 +42,71 @@ class RobotSelectEntityDescription(
+            select_fn=lambda robot, option: robot.set_night_light_mode(
+                NightLightMode[option.upper()]
+            ),
+            icon_fn=lambda robot: "mdi:lightbulb-question"
+            if (mode := robot.night_light_mode) is None
+            else "mdi:lightbulb-auto"
+            if mode == NightLightMode.AUTO
+            else "mdi:lightbulb-on"
+            if mode == NightLightMode.ON
+            else "mdi:lightbulb-off",
```
Same is above

tkdrob commented on tests/components/litterrobot/test_select.py:
```diff
@@ -63,3 +65,18 @@ async def test_invalid_wait_time_select(hass: HomeAssistant, mock_account):
+
+    state = hass.states.get("select.test_night_light_level")
+    assert state.state == "Low"
+    assert state.attributes.get(ATTR_ICON) == "mdi:lightbulb-on-30"
```
I'm not sure if we need to test the icons as they are aesthetic only. Perhaps we should should assert the possible options instead?

natekspencer replied:
Even though it's aesthetic, it's still code, so this ensures logic and coverage. But it is probably wise to add the options as a check as well.

tkdrob replied:
I like that answer and take back the first part. It's prone to errors so yeah. It seems best to keep the assertion.

tkdrob commented on tests/components/litterrobot/test_select.py:
```diff
@@ -63,3 +65,18 @@ async def test_invalid_wait_time_select(hass: HomeAssistant, mock_account):
+
+    state = hass.states.get("select.test_night_light_mode")
+    assert state.state == "Off"
+    assert state.attributes.get(ATTR_ICON) == "mdi:lightbulb-off"
```
Same as above

natekspencer committed:
Use a map for icons

natekspencer committed:
Add options check to tests

cdce8p committed:
Add strict typing for radarr (#79242)

cdce8p committed:
Add strict typing for lidarr (#79241)

cdce8p committed:
Add strict typing for WLED (#79822)

* Add strict typing for WLED

* Update backoff constraint

bouwew committed:
Bump plugwise to v0.21.4 (#79831)

cereal2nd committed:
Bumb velbusaio to 2022.10.3 (#79821)

* Bumb velbusaio to 2022.10.3

* Handle the possibility that get_cover position is None (unknown) in previous versions this was always 0

HarvsG committed:
Improve typing and code quality in beyesian (#79603)

* strict typing

* Detail implication

* adds newline

* don't change indenting

* really dont change indenting

* Update homeassistant/components/bayesian/binary_sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* typing in async_setup_platform() + remove arg

* less ambiguity

* mypy thinks Literal[False] otherwise

* clearer log

* don't use `and` assignments

* observations not values

* clarify can be None

* observation can't be none

* assert we have at least one

* make it clearer where we're using UUIDs

* remove unnecessary bool

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* Unnecessary None handling

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* Better type setting

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* Reccomended changes.

* remove if statement not needed

* Not strict until _TrackTemplateResultInfo fixed

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

bdraco committed:
Bump dbus-fast to 1.29.1 (#79787)

* Bump dbus-fast to 1.28.0

Performance improvements

changelog: https://github.com/Bluetooth-Devices/dbus-fast/compare/v1.26.0...v1.28.0

* bump again

* bump for cleanups

jeinwag committed:
Add device trigger for Lidl Silvercrest switch to deCONZ (#79839)

* add deconz support for Lidl Silvercrest switch

* Update homeassistant/components/deconz/device_trigger.py

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

* Update homeassistant/components/deconz/device_trigger.py

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

* clarify it's a button, remove turn on event

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

tkdrob committed:
Move Sonarr API calls to coordinators (#79826)

lennart-k committed:
Fix realtime option for hvv_departures (#79799)

tkdrob committed:
Add strict typing to Google Sheets (#79801)

bdr99 committed:
Add Mazda brand (#79683)

philschenker committed:
Change shelly trv precision to what is supported (#79672)

change shelly trv precision to what is supported

Shelly TRVs do support half-degree steps, change this accordingly.

bieniu committed:
Use new device classes in Accuweather integration (#79717)

* Add new device classes

* Update tests

G-Two committed:
Add vehicle model/year to subaru device (#79484)

HennieLP committed:
Add state class to bosch_shc energy sensor (#79470)

Make That energy sensor works in Dashbord

Commit:
[ci skip] Translation update

tkdrob committed:
Migrate Sonarr to new entity naming style (#79844)

spycle committed:
Fix keymitt_ble discovery (#79809)

* Fix keymitt_ble discovery

* Update tests

* Up version

* Up version keymitt_ble

* Up version keymitt_ble

Kane610 committed:
Fix POE control port_idx error in UniFi (#79838)

Bump UniFi dependency

cdce8p committed:
Use value_fn in WLED number (#79865)

bertmelis committed:
Process abbreviated availability options in mqtt discovery payload (#79712)

Expand availability in mqtt discovery payload

gjohansson-ST committed:
Bump pytrafikverket to 0.2.1 (#79872)

cdce8p committed:
Update typing-extensions constraint to >=4.4.0 (#79860)

Nardol committed:
Add state class measurement to SwitchBot signal strength sensors (#79886)

zhibek committed:
Fix london_underground TUBE_LINES to match current API output (#79410)

Fix: Update london_underground component with updated TUBE_LINES list to match current API output

scop committed:
Add syncthru active alerts sensor, set default manufacturer (#79418)

* Use Samsung as default manufacturer

* Sensor docstring fixes

* Add active alerts sensor

Danielhiversen committed:
Correct unit for Opengarage rssi sensor (#79403)

Signed-off-by: Daniel Hjelseth Høyer <github@dahoiv.net>

Signed-off-by: Daniel Hjelseth Høyer <github@dahoiv.net>

starkillerOG committed:
Add Roborock as supported brand of xiaomi miio (#79312)

* Add Roborock as supported brand

* Update supported_brands.py

starkillerOG committed:
Add supported brands for Motion Blinds  (#79301)

* Add ScreenAway

* Add aditional brands

majuss committed:
Bump lupupy to support XT2 and up (#79289)

* Bumped lupupy to support XT2 and up

* requirements script

thecode committed:
Migrate Shelly to use kelvin for color temperature (#79880)

bieniu committed:
Rework Brother sensor platform (#79864)

* Rework BrotherSensorEntityDescription

* Rework state attributes

* Cleaning

* Add _handle_coordinator_update()

* Suggested change

* Re-add consts

puddly committed:
Bump ZHA dependencies (#79898)

bachya committed:
Remove redundant OpenUV test fixture (#79905)

bachya committed:
Remove redundant Ridwell test fixture (#79906)

bachya committed:
Remove redundant ReCollect Waste test fixture (#79907)

bachya committed:
Remove redundant WattTime test fixture (#79909)

bachya committed:
Remove redundant Notion test fixture (#79910)

bachya committed:
Remove redundant IQVIA test fixture (#79911)

bachya committed:
Remove redundant Tile test fixture (#79912)

bachya committed:
Move AirNow test fixtures to `conftest.py` (#79902)

* Move AirNow test fixtures to `conftest.py`

* Unnecessary fixture

* Better

* Linting

iMicknl committed:
Update Config Flow to show message about unsupported Overkiz hardware (#79503)

* Update Config Flow to show cozytouch unsupported hardware error

* Apply feedback

* Remove vague unknown user exception

* Fix test

* Code coverage back to 100%

cgtobi committed:
Bump pyatmo to 7.1.1 (#79918)

Commit:
[ci skip] Translation update

bdraco committed:
Bump dbus-fast to 1.33.0 (#79921)

ctalkington committed:
Use persistent device id for jellyfin requests (#79840)

jbouwh committed:
Correct use of ConfigType in MQTT config flow code (#79934)

Correct use of ConfigType

bieniu committed:
Migrate attributes to separate sensors in Brother integration (#79932)

Migrate attributes to sensors

gjohansson-ST committed:
Remove yaml import openexchangerates (#79856)

* Depr openexchangerates yaml

* Remove setup_platform and issue

* Remove schema

* Remove not longer used constant

TheJulianJES committed:
Add friendly name to ZHA config entities (#79926)

* Add friendly name to ZHA config entities

* Follow HA capitalization conventions

* Change from "Start-up level" to "Start-up current level"

* Remove siren select friendly names (temporarily)

* Change tests to expect new entity ids

* Re-add siren select friendly names

* Change siren tests to expect new entity ids

scop committed:
Remove incorrect UpCloud logger from manifest (#79855)

HarvsG committed:
Make _TrackTemplateResultInfo not private (#79812)

gjohansson-ST committed:
Remove not used string from openexchangerates (#79937)

HarvsG committed:
Enable strict typing on Bayesian (#79870)

* make bayesian static

* no longer private

jeeftor committed:
Add configuration URL to IPP (Printer) (#79313)

switch to 0.12.0 ipp lib

danaues committed:
Add support for parent_device field so entities are nested within Keypad Devices (#79513)

Co-authored-by: J. Nick Koston <nick@koston.org>

frenck committed:
Add sensor platform to LaMetric (#79935)

danaues committed:
Add support for Homeowner and Phantom Keypads (#79958)

Djelibeybi committed:
Migrate the LIFX integration to use kelvin for color temp (#79775)

bdraco committed:
Bump dbus-fast to 1.38.0 (#79962)

ctalkington committed:
Bump jellyfin-apiclient-python to 1.9.2 (#79945)

Commit:
[ci skip] Translation update

cgarwood committed:
Fix typo SIGNAL_BOOTSTRAP_INTEGRATONS -> SIGNAL_BOOTSTRAP_INTEGRATIONS (#79970)

bdraco committed:
Bump bluetooth-auto-recovery to 0.3.4 (#79971)

ctalkington committed:
Use server name as entry title in Jellyfin (#79965)

ctalkington committed:
Add tests for Jellyfin init (#79968)

frenck committed:
Add select platform to LaMetric (#79803)

dependabot[bot] committed:
Bump actions/stale from 6.0.0 to 6.0.1 (#79977)

Bumps [actions/stale](https://github.com/actions/stale) from 6.0.0 to 6.0.1.
- [Release notes](https://github.com/actions/stale/releases)
- [Changelog](https://github.com/actions/stale/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/stale/compare/v6.0.0...v6.0.1)

---
updated-dependencies:
- dependency-name: actions/stale
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

frenck committed:
Remove system marker from Supervisor integration (#79997)

epenet committed:
Add docstring to Sensor enums (#79983)

* Add docstring to Sensor enums

* Adjust MONETARY docstring

molisani committed:
Update to pygtfs 0.1.7 (#79975)

cgarwood committed:
Add load_url service to fully_kiosk integration (#79969)

flemmingss committed:
Clean duplicate nextcloud sensor (#79900)

Update __init__.py

Removed duplicate of  "nextcloud_database_version"

Nardol committed:
Migrate Switchbot to new entity naming style (#80008)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

frenck committed:
Update black to 22.10.0 (#80006)

epenet committed:
Add options to SelectEntityDescription (#78882)

frenck committed:
Update apprise to 1.1.0 (#80009)

emontnemery committed:
Adapt group to color temperature in K (#79719)

* Adapt group to color temperature in K

* Adjust tests

* Adjust tests

emontnemery committed:
Correct min/max mireds for lights which use K for color temp (#79998)

rappenze committed:
Fix armed extra state attribute in fibaro entity (#80034)

epenet committed:
Adjust device classes in smartthings (#79283)

Commit:
Bump pyhiveapi to 0.5.14 (#79530)

wrt54g committed:
Update screenshot (#79459)

* Delete old image

* Add new screenshot

* New image

Jc2k committed:
Fix Eve Thermo always showing as heating in homekit_controller even when off (#80019)

Djelibeybi committed:
Add support for the Flame and Morph effects for Tile and Candle (#80014)

kingy444 committed:
Powerview refactor prep for all shade types (#79862)

Kane610 committed:
Adapt deCONZ binary sensors to entity descriptions (#79486)

Now typing with lambdas work

bramkragten committed:
Update frontend to 20221010.0 (#79994)

dmulcahey committed:
Add friendly entity names for ZHA sensors (#80035)

* Add friendly entity names for ZHA sensors

* lowercase 2nd word

emontnemery committed:
Disable echo for non SQLite databases (#80032)

* Disable echo for non SQLite databases

* Add test
